### PR TITLE
Make public addresses go first in authority discovery DHT records

### PR DESCRIPTION
**Pull request by dmitry-markin**
Make sure explicitly set by the operator public addresses go first in the authority discovery DHT records.

Also update `Discovery` behavior to eliminate duplicates in the returned addresses.

This PR should improve situation with https://github.com/paritytech/polkadot-sdk/issues/3519.

Obsoletes https://github.com/paritytech/polkadot-sdk/pull/3657.

### CODE REVIEW

File: substrate/client/authority-discovery/src/worker.rs [hunk 1]
```diff
@@ -304,32 +338,48 @@ where
...
-				multiaddr::Protocol::Ip6(ip) if !IpNetwork::from(ip).is_global() => false,
-				_ => true,
+				if self.public_addresses.contains(&address) {
+					// Already added above.
+					None
+				} else {
+					Some(address)
+				}
+			}))
+			.filter(move |address| {
```
Comment (dmitry-markin) [inline substrate/client/authority-discovery/src/worker.rs:359]: Not sure, may be we should always publish public addresses even if they are non-global.

**Review by lexnv — APPROVED** [review]
LGTM! Nice one 👍

File: substrate/client/authority-discovery/src/worker.rs (hunk 1 above)
Comment (dmitry-markin) [reply substrate/client/authority-discovery/src/worker.rs:359]: I was thinking about a testnet in a private network case. But we can also use `public_non_global_ips` flag for this.

Comment (bkchr) [reply substrate/client/authority-discovery/src/worker.rs:359]: Yeah private network. When you run it locally on your computer, you also need authority discovery to work ;)

**Review by bkchr — APPROVED** [review]
Ty!

One thing I'm not sure if we are already doing it, could we ensure that `external_addresses` contains the listen addresses? Maybe we could even have an extra function for getting the listen addresses to have in the end, [public addresses, listen addresses, external addresses].

**Comment by dmitry-markin** [comment]
> One thing I'm not sure if we are already doing it, could we ensure that `external_addresses` contains the listen addresses? 

`external_addresses` do not contain listen addresses, and I thought it's a feature and not a bug.

> Maybe we could even have an extra function for getting the listen addresses to have in the end, [public addresses, listen addresses, external addresses].

Do you mean [public addresses, external addresses, listen addresses]?

**Comment by bkchr** [comment]
> `external_addresses` do not contain listen addresses, and I thought it's a feature and not a bug.

IDK :D I mean this was more some proposal by me. Also should not block this pr BTW.



> Do you mean [public addresses, external addresses, listen addresses]?

Listen addresses are more "trusted" than external addresses as you can just query your local network devices to get their ip addresses etc. For validators which should be exposed to the internet, the listen addresses should also give you the correct ip addresses directly. If the operator is giving public addresses, I assume that listen and public will probably be the same.

**Comment by dmitry-markin** [comment]
> > `external_addresses` do not contain listen addresses, and I thought it's a feature and not a bug.
> 
> IDK :D I mean this was more some proposal by me. Also should not block this pr BTW.
> 
> > Do you mean [public addresses, external addresses, listen addresses]?
> 
> Listen addresses are more "trusted" than external addresses as you can just query your local network devices to get their ip addresses etc. For validators which should be exposed to the internet, the listen addresses should also give you the correct ip addresses directly. If the operator is giving public addresses, I assume that listen and public will probably be the same.

We can address it in a follow-up PR. Personally, I would refrain from adding listen addresses as they contain all the interfaces (including 127.0.0.1 and private networks), and it should be enough to have just one valid Identify response from a remote peer to learn a valid external address. We should take into account address translation done by libp2p (it assign a listen address port to discovered external addresses) to make an informed decision though.

**Pull request merged**